Spanish translation for section async (#1290)
fastapi/fastapi: docs/es/docs/async.md

Para entender las diferencias, imagine la siguiente historia sobre hamburguesas:

### Hamburguesas Concurrentes

Vas con tu pareja a pedir comida rápida, haces cola mientras el cajero recoge los pedidos de las personas de delante tuyo.

Te llega tu turno, haces tu pedido de 2 hamburguesas impresionantes para tu pareja y para ti.

Pagas.

El cajero le dice algo al chico de la cocina para que sepa que tiene que preparar tus hamburguesas (a pesar de que actualmente está preparando las de los clientes anteriores).

El cajero te da el número de tu turno.

Mientras esperas, vas con tu pareja y eliges una mesa, os sentáis y habláis durante un rato largo  (ya que las hamburguesas son muy impresionantes y llevan un rato prepararse).

Mientras estás  sentado en la mesa con tu  pareja, esperando las hamburguesas, puedes disfrutar ese tiempo admirando lo increíble, guap@ e inteligente que es tu pareja.

Mientras esperas y hablas con tu pareja, de vez en cuando, verificas el número del mostrador para ver si ya es tu turno.

Al final, en algún momento, llega tu turno. Vas al mostrador, coges tus hamburguesas y vuelves a la mesa.

> Tú y tu pareja os coméis las hamburguesas y os divertís.

Author: Not sure about this one but... ;)

Eso, más el simple hecho de que Python es el lenguaje principal para **Data Science**, Machine Learning y especialmente Deep Learning, hacen de FastAPI una muy buena combinación para las API y aplicaciones web de Data Science / Machine Learning (entre muchas otras).

Para ver cómo lograr este paralelismo en producción, consulte la sección sobre [Implementación] (deployment.md) {. Internal-link target = _blank}.

## `async` y` await`

Las versiones modernas de python tienen una forma muy intuitiva de definir código asíncrono. Esto hace que se vea como un código "secuencial" normal y que haga la "espera" por usted en los momentos correctos.

Cuando hay una operación que requerirá esperar antes de dar los resultados y tiene soporte para estas nuevas características de Python, puede codificarlo como:

```Python
burgers = await get_burgers(2)
```

La clave aquí es la 'espera'. Le dice a Python que tiene que esperar a que `get_burgers (2)` termine de hacer lo suyo antes de almacenar los resultados en `hamburguesas`. Con eso, Python sabrá que puede ir y hacer otra cosa mientras tanto (como recibir otra solicitud).

> Para que `await` funcione, tiene que estar dentro de una función que admita esta asincronía. Para hacer eso, simplemente lo declara con `async def`:

Reviewer: ```suggestion
Para que `await` funcione, tiene que estar dentro de una función que admita esta asincronía. Para hacer eso, simplemente lo declaras con `async def`:
```

```Python hl_lines="1"
async def get_burgers(number: int):
    # Do some asynchronous stuff to create the burgers
    return burgers
```

...instead of `def`:

```Python hl_lines="2"
# This is not asynchronous
def get_sequential_burgers(number: int):
    # Do some sequential stuff to create the burgers
    return burgers
```

> Con `async def`, Python sabe que, dentro de esa función, debe tener en cuenta las expresiones` wait ', y que puede "pausar" la ejecución de esa función e ir a hacer otra cosa antes de regresar.

Reviewer: ```suggestion
Con `async def`, Python sabe que, dentro de esa función, debe tener en cuenta las expresiones `wait` y que puede "pausar" la ejecución de esa función e ir a hacer otra cosa antes de regresar.
```

> Cuando desee llamar a una función `async def`, debe" esperarla ". Entonces, esto no funcionará:

Reviewer: ```suggestion
Cuando desees llamar a una función `async def`, debes" esperarla ". Entonces, esto no funcionará:
```

```Python
# Esto no funcionará, porque get_burgers se definió con: async def
hamburguesas = get_burgers (2)
```

---

> Por lo tanto, si está utilizando una biblioteca que le dice que puede llamarla con `await`, debe crear las *funciones de enrutado* que la usan con` async def`, como en:

Reviewer: ```suggestion
Por lo tanto, si estás utilizando una biblioteca que te dice que puedes llamarla con `await`, debes crear las *path operation functions* que la usan con `async def`, como en:
```

```Python hl_lines="2 3"
@app.get('/burgers')
async def read_burgers():
    burgers = await get_burgers(2)
    return burgers
```

### Más detalles técnicos

> Es posible que haya notado que 'esperar' solo se puede usar dentro de las funciones definidas con 'async def'.

Reviewer: ```suggestion
Es posible que hayas notado que `await` solo se puede usar dentro de las funciones definidas con `async def`.
```

> Pero al mismo tiempo, las funciones definidas con `async def` deben ser "esperadas". Por lo tanto, las funciones con `async def` solo se pueden invocar dentro de las funciones definidas con` async def` también.

Reviewer: ```suggestion
Pero al mismo tiempo, las funciones definidas con `async def` deben ser "esperadas". Por lo tanto, las funciones con `async def` solo se pueden invocar dentro de las funciones definidas con `async def` también.
```

Entonces, relacionado con la paradoja del huevo y la gallina, ¿cómo se llama a la primera función `async`?

> Si está trabajando con **FastAPI** no tiene que preocuparse por eso, porque esa "primera" función será su *función de enrutados*, y FastAPI sabrá cómo hacer lo pertinente.

Reviewer: ```suggestion
Si estás trabajando con **FastAPI** no tienes que preocuparte por eso, porque esa "primera" función será tu *path operation function*, y FastAPI sabrá cómo hacer lo pertinente.
```

> En el caso de que desee usar `async` /` await` sin FastAPI, <a href="https://docs.python.org/3/library/asyncio-task.html#coroutine" class="external-link" target="_blank">check the official Python docs</a>.

Reviewer: ```suggestion
En el caso de que desees usar `async` / `await` sin FastAPI, <a href="https://docs.python.org/3/library/asyncio-task.html#coroutine" class="external-link" target="_blank">revisa la documentación oficial de Python</a>.
```

### Otras formas de código asíncrono

> Este estilo de usar `async` y` await` es relativamente nuevo en el lenguaje.

Reviewer: ```suggestion
Este estilo de usar `async` y `await` es relativamente nuevo en el lenguaje.
```

Pero hace que trabajar con código asíncrono sea mucho más fácil.

Esta misma sintaxis (o casi idéntica) también se incluyó recientemente en las versiones modernas de JavaScript (en Browser y NodeJS).

Pero antes de eso, manejar código asíncrono era bastante más complejo y difícil.

> En versiones anteriores de Python, podría haber utilizado hilos o <a href="http://www.gevent.org/" class="external-link" target="_blank">Gevent</a>. Pero el código es mucho más complejo de entender, depurar y desarrollar.

Reviewer: ```suggestion
En versiones anteriores de Python, podrías haber utilizado hilos o <a href="http://www.gevent.org/" class="external-link" target="_blank">Gevent</a>. Pero el código es mucho más complejo de entender, depurar y desarrollar.
```

> En versiones anteriores de NodeJS / Browser JavaScript, habría utilizado "callbacks". Lo que conduce a <a href="http://callbackhell.com/" class="external-link" target="_blank"> callback hell </a>.

Reviewer: ```suggestion
En versiones anteriores de NodeJS / Browser JavaScript, habrías utilizado "callbacks". Lo que conduce a <a href="http://callbackhell.com/" class="external-link" target="_blank">callback hell</a>.
```

> ## Corutinas

Reviewer: ```suggestion
## Coroutines
```

> **Coroutine** es el término guay para la cosa devuelta por una función `async def`. Python sabe que es algo así como una función que puede iniciar y que terminará en algún momento, pero que también podría pausarse internamente, siempre que haya una `await` dentro de ella.

Reviewer: ```suggestion
**Coroutine** es un término sofisticado para referirse a la cosa devuelta por una función `async def`. Python sabe que es algo así como una función que puede iniciar y que terminará en algún momento, pero que también podría pausarse internamente, siempre que haya una `await` dentro de ella.
```

> Pero toda esta funcionalidad de usar código asincrónico con `async` y` await` se resume muchas veces como el uso de "corutinas". Es comparable a la característica principal de Go, las "Goroutines".

Reviewer: ```suggestion
Pero toda esta funcionalidad de usar código asincrónico con `async` y `await` se resume muchas veces como el uso de "coroutines". Es comparable a la característica principal de Go, las "Goroutines".
```

## Conclusión

Veamos la misma frase de arriba:

> > Las versiones modernas de Python tienen soporte para **"código asíncrono"** usando algo llamado **"corutinas"**, con la sintaxis **`async` y` await`**.

Reviewer: ```suggestion
> Las versiones modernas de Python tienen soporte para **"código asíncrono"** usando algo llamado **"coroutines"**, con la sintaxis **`async` y `await`**.
```

Eso ya debería tener más sentido ahora.

Todo eso es lo que impulsa FastAPI (a través de Starlette) y lo que lo hace tener un rendimiento tan impresionante.

## Detalles muy técnicos

!!! advertencia
    Probablemente puedas saltarte esto.

    Estos son detalles muy técnicos de cómo **FastAPI** funciona a muy bajo nivel.
    

>     Si tienes bastante conocimiento técnico (corutinas, hilos, bloqueos, etc.) y tienes curiosidad acerca de cómo FastAPI gestiona `async def` vs normal `def`, continúa.

Reviewer: ```suggestion
    Si tienes bastante conocimiento técnico (coroutines, threads, bloqueos, etc.) y tienes curiosidad acerca de cómo FastAPI gestiona `async def` vs `def` normal, continúa.
```

> ### Funciones de enrutado 

Reviewer: ```suggestion
### Path operation functions 
```

> Cuando declara una *función de enrutado* con `def` normal en lugar de `async def`, se ejecuta en un conjunto de subprocesos externo que luego  espera, en lugar de ser llamado directamente (ya que bloquearía el servidor).

Reviewer: ```suggestion
Cuando declaras una *path operation function* con `def` normal en lugar de `async def`, se ejecuta en un threadpool externo que luego espera, en lugar de ser llamado directamente (ya que bloquearía el servidor).
```

> Si vienes de otro framework asíncrono que no funciona de la manera descrita anteriormente y estás acostumbrado a definir funciones de enrutado *del tipo sólo cálculo* con `def` simple para una pequeña ganancia de rendimiento (aproximadamente 100 nanosegundos), tenga en cuenta que en **FastAPI** el efecto sería bastante opuesto. En estos casos, es mejor usar `async def` a menos que sus *funciones de enrutado* usen un código que realice el bloqueo <abbr title="Input/Output: disk reading or writing, network communications.">IO</abbr>.

Reviewer: ```suggestion
Si vienes de otro framework asíncrono que no funciona de la manera descrita anteriormente y estás acostumbrado a definir *path operation functions* del tipo sólo cálculo con `def` simple para una pequeña ganancia de rendimiento (aproximadamente 100 nanosegundos), ten en cuenta que en **FastAPI** el efecto sería bastante opuesto. En estos casos, es mejor usar `async def` a menos que tus *path operation functions* usen un código que realice el bloqueo <abbr title="Input/Output: disk reading or writing, network communications.">IO</abbr>.
```

> Aún así, en ambas situaciones, es probable que **FastAPI** sea [aún más rápido](/#rendimiento) {.Internal-link target=_blank} que (o al menos comparable) a su framework anterior.

Reviewer: ```suggestion
Aún así, en ambas situaciones, es probable que **FastAPI** sea [aún más rápido](/#rendimiento){.Internal-link target=_blank} que (o al menos comparable) a tu framework anterior.
```

### Dependencias

> Lo mismo se aplica para las dependencias. Si una dependencia es una función estándar `def` en lugar de` async def`, se ejecuta en el conjunto de hilos externo.

Reviewer: ```suggestion
Lo mismo se aplica para las dependencias. Si una dependencia es una función estándar `def` en lugar de `async def`, se ejecuta en el threadpool externo.
```

### Subdependencias

> Puede tener múltiples dependencias y subdependencias que se requieren entre sí (como parámetros de las definiciones de función), algunas de ellas pueden crearse con `async def` y otras con` def` normal. Seguiría funcionando correctamente, y los creados con `def` normal se llamarían en un hilo externo en lugar de ser " awaited ".

Reviewer: ```suggestion
Puedes tener múltiples dependencias y subdependencias que se requieren entre sí (como parámetros de las definiciones de función), algunas de ellas pueden crearse con `async def` y otras con `def` normal. Seguiría funcionando correctamente, y los creados con `def` normal se llamarían en un thread externo en lugar de ser "awaited".
```

### Otras "utility functions"

> Cualquier otra función de utilidad que llame directamente se puede crear con `def` o` async def` normales y FastAPI no afectará la manera en que la llama.

Reviewer: ```suggestion
Cualquier otra función de utilidad que llame directamente se puede crear con `def` o `async def` normales y FastAPI no afectará la manera en que la llamas.
```

> Esto choca con las funciones que FastAPI le solicita: *funciones de enrutado* y dependencias.

Reviewer: ```suggestion
Esto contrasta con las funciones que FastAPI llama por ti: *path operation functions* y dependencias.
```

> Si su utility function es una función normal con `def`, se llamará directamente (tal cual la escribes en su código), no en un conjunto de hilos, si la función se crea con` async def`, entonces debes usar await en  esa función cuando la llamas en tu código.

Reviewer: ```suggestion
Si tu utility function es una función normal con `def`, se llamará directamente (tal cual la escribes en tu código), no en un threadpool, si la función se crea con `async def`, entonces debes usar await en esa función cuando la llamas en tu código.
```
